### PR TITLE
Add flush() to all supported resource classes

### PR DESCRIPTION
**Pull request by KM4YRI**

### CODE REVIEW

File: pyvisa/resources/tcpip.py [hunk 1]
```diff
@@ -39,3 +39,12 @@ class TCPIPSocket(MessageBasedResource):
 
     Do not instantiate directly, use :meth:`pyvisa.highlevel.ResourceManager.open_resource`.
     """
+
+    def flush(self, mask):
```
Comment (MatthieuDartiailh) [inline pyvisa/resources/tcpip.py]: This won't be needed if we implement above comment.

File: pyvisa/resources/gpib.py [hunk 2]
```diff
@@ -146,3 +146,12 @@ def group_execute_trigger(self, *resources):
         command.append(0x08)
 
         return self.send_command(bytes(command))
+
+    def flush(self, mask):
```
Comment (MatthieuDartiailh) [inline pyvisa/resources/gpib.py]: This won't be needed if we follow below comment.

**Comment by MatthieuDartiailh** [comment]
ping @KM4YRI

**Comment by MatthieuDartiailh** [comment]
ping @KM4YRI

**Comment by KM4YRI** [comment]
Sorry, I've been slammed at work and will be for at least the next few weeks.  Feel free to either keep the issue and PR open and I'll get to it when I can, or close them if you'd like.  Sorry for flaking out on this.

**Comment by MatthieuDartiailh** [comment]
No Problem @KM4YRI. I will see if I can find time to address my comments and test in that case. I will ask you to review my changes, if that is alright with you, so that we have at least a second pair of eyes on them before merging.

**Comment by MatthieuDartiailh** [comment]
Checking again I cannot find a single reference suggesting that this is not supported on USB... No idea why I said this in the first place.

**Comment by MatthieuDartiailh** [comment]
I updated the docstring of flush based on the IVI standard and removed the method from the SOCKET resource class which inherits from MessageBasedResource. If anybody can test, it would be great.

**Comment by MatthieuDartiailh** [comment]
@KM4YRI if you can review and test it would be great.

**Comment by MatthieuDartiailh** [comment]
@KM4YRI I would like to merge this soon for it to be part of the next release (either next week or at the beginning of August). Could you give it a last look ?

**Comment by ceres-c** [comment]
Hello, sorry if I bumping this thread, but I wanted to make sure that I understood the current status of ```flush```. It is now exposed by all the resource types, but the VISA library function is not actually exposed via the python ctypes bindings, am I right?

New to this lib, but I assume the implementation should be in ```highlevel.VisaLibraryBase```, but that is just raising a ```NotImplementedError```. Did I miss something out, or is it not possible to clear the TCP buffer?

**Comment by MatthieuDartiailh** [comment]
Flush is supported by all message based resources without exceptions. It is exposed by the ctypes wrapper. The class you refer to is the abstract base class that needs to be implemented by all backends.

**Comment by ceres-c** [comment]
Thanks, then I thought I had it right at first, but I was puzzled as I actually got some `NotImplementedError`.

Now that I think of it it might have been that I was debugging the script, and maybe that was it. I'll try again when I'm at the lab on Monday, thanks

**Comment by ceres-c** [comment]
Am I retarded?
```python3
self.rm = pyvisa.ResourceManager()
self.sc = self.rm.open_resource(f'TCPIP::{ip}::INSTR') # type: ignore
print(self.sc.query('*IDN?'))
self.sc.flush(pyvisa.constants.VI_READ_BUF_DISCARD | pyvisa.constants.VI_WRITE_BUF_DISCARD)
```
Result
```
TEKTRONIX,MSO44,C019654,CF:91.1CT FV:2.0.3.950

Traceback (most recent call last):
  File "/home/federico/tools/chipwhisperer/user/mso44_test.py", line 5, in <module>
    mso44: cw.scopes.MSO4 = cw.scope(
                            ^^^^^^^^^
  File "/home/federico/tools/chipwhisperer/cw-devel/lib/python3.11/site-packages/chipwhisperer/__init__.py", line 388, in scope
    rtn.con(**kwargs)
  File "/home/federico/tools/chipwhisperer/cw-devel/lib/python3.11/site-packages/chipwhisperer/capture/scopes/MSO4.py", line 114, in con
    self.sc.flush(pyvisa.constants.VI_READ_BUF_DISCARD | pyvisa.constants.VI_WRITE_BUF_DISCARD)
  File "/home/federico/tools/chipwhisperer/cw-devel/lib/python3.11/site-packages/pyvisa/resources/messagebased.py", line 797, in flush
    self.visalib.flush(self.session, mask)
  File "/home/federico/tools/chipwhisperer/cw-devel/lib/python3.11/site-packages/pyvisa_py/highlevel.py", line 222, in flush
    return self.handle_return_value(session, sess.flush(mask))
                                             ^^^^^^^^^^^^^^^^
  File "/home/federico/tools/chipwhisperer/cw-devel/lib/python3.11/site-packages/pyvisa_py/sessions.py", line 435, in flush
    raise NotImplementedError
```

Every other operation seems to be working fine on the same object

**Comment by arr-ee** [comment]
Please avoid using such language.

Your issue is not with `pyvisa` but with `pyvisa-py`, which, in fact, does not implement `flush` for VXI-11 (`TCPIP::...::INSTR`), only for raw sockets (`TCPIP::...::SOCKET`): https://github.com/pyvisa/pyvisa-py/blob/a9221593a2b254ed55a05c7525240eee16c6e46d/pyvisa_py/tcpip.py#L1273

Feel free to open an issue in that project, ideally with a testcase highlighting the need for explicit flush.

**Comment by ceres-c** [comment]
Sorry, but I actually felt really dumb, and indeed, I was right in that. Will ask in the right place

**Pull request merged**